File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque

global M, N, H
M, N, H = map(int, input().split())

arr = []
for _ in range(H):
    brr = []
    for _ in range(N):
        brr.append(list(map(int, input().split())))
    arr.append(brr)

global queue
queue = deque()

def putq(x, y, z, day):
    global M, N, H, queue
    if 0 <= x < M and 0 <= y < N and 0 <= z < H:
        queue.append([x, y, z, day])

nearx = [1, -1, 0, 0, 0, 0]
neary = [0, 0, 1, -1, 0, 0]
nearz = [0, 0, 0, 0, 1, -1]

ans = 0

# BFS
# 모든 배열에 접근
for z in range(H):
    for y in range(N):
        for x in range(M):
            # 만약 1이라면
            if arr[z][y][x] == 1:
                # 주변 6개를 queue에 넣음
                for i in range(6):
                    # day = 0
                    putq(x+nearx[i], y+neary[i], z+nearz[i], 0)


# queue가 비어있지 않을 때까지
while queue:
    # print(queue)
    # queue에서 pop
    item = queue.popleft()
    x = item[0]
    y = item[1]
    z = item[2]
    day = item[3]
    # 이 때 day가 진짜. ans = max(res, day)
    ans = max(ans, day)
    # 만약 0이라면
    if arr[z][y][x] == 0:
        arr[z][y][x] = 1
        # 주변 6개 모두 queue에 넣음 / day +1
        for i in range(6):
            putq(x + nearx[i], y + neary[i], z + nearz[i], day+1)
    # 만약 0이 아니라면
    else:
        continue

allsame = True

# print(arr)

# 모든 배열에 접근
for z in range(H):
    for y in range(N):
        for x in range(M):
            # 모든 토마토가 1이면 res 출력
            if arr[z][y][x] == 0:
                allsame = False
                print(-1)
                exit()
# 아니라면 -1 출력
print(ans)

File: 백준/Gold/7576. 토마토/토마토.swift
import Foundation

let input = readLine()!.split(separator: " ").map{Int(String($0))!}
let M = input[0]
let N = input[1]

var visit: [(Int, Int, Int)] = []
var res: Int = 0

var arr: [[Int]] = []

for _ in 0..<N{
    arr.append(readLine()!.split(separator: " ").map{Int(String($0))!})
}


for y in 0..<N{
    for x in 0..<M{
        if arr[y][x] == 1{
            visit.append((x, y, 0))
        }
    }
}

let xx = [0, 0, -1, 1]
let yy = [-1, 1, 0, 0]
var id = 0

while id < visit.count{
    let item = visit[id]
    id += 1
    
    let x = item.0
    let y = item.1
    var day = item.2
    res = max(res, day)

    for idx in 0...3{
        let nx = x+xx[idx]
        let ny = y+yy[idx]
        if nx >= 0 && nx < M && ny >= 0 && ny < N{
            if arr[ny][nx] == 0{
                visit.append((nx, ny, day+1))
                arr[ny][nx] = 1
//                print(visit)
//                for item in arr{
//                    print(item)
//                }
            }
        }
    }
}

for temp in arr{
    for item in temp{
        if item == 0{
            print(-1)
            exit(0)
        }
    }
}

print(res)
